packages/go_router/lib/src/route.dart

```diff
@@ -57,6 +59,12 @@ typedef StatefulShellRoutePageBuilder = Page<dynamic> Function(
 typedef NavigatorBuilder = Widget Function(
     List<NavigatorObserver>? observers, String? restorationScopeId);
 
+/// Signature for function used in [RouteBase.onExit].
+///
+/// If the return value is true or the future resolve to true, the route will
+/// exit as usual. Otherwise, the operation will abort.
+typedef ExitCallback = FutureOr<bool> Function(BuildContext context);
```

Comment:
Should we use `Future<bool>` instead? Using `FutureOr<T>` is discouraged iirc.

I get the intent, but a user can always do `onExit: (BuildContext context) async => false`

Reply:
We will like to provide a chance to handle this synchronously. Otherwise it would introduce one frame delay between navigation. Also the ship has sailed for GoRoute.redirect, so I will keep this as is.